### PR TITLE
DownloadResolver specifies hash type from config

### PR DESCRIPTION
**Pull request by JuanCarlosGI**

### CODE REVIEW

**Comment by dannyvv** [comment]
@imnarasa@micorosft.com: Can you have a look too?

File: Public/Src/Utilities/Storage/ContentHashingUtilities.cs [hunk 1]
```diff
@@ -87,6 +105,8 @@ public static void DisposeAndResetHasher()
         {
             s_hasher?.Dispose();
             s_hasher = HashInfo.CreateContentHasher();
+
+            s_contentHasherByHashType = new ConcurrentDictionary<HashType, IContentHasher>();
```
Comment (dannyvv) [inline Public/Src/Utilities/Storage/ContentHashingUtilities.cs]: shoudl we dispose the ones in the dictionary

File: Public/Src/Utilities/Storage/ContentHashingUtilities.cs [hunk 2]
```diff
@@ -22,6 +23,23 @@ public static class ContentHashingUtilities
...
+            {
+                return s_hasher;
+            }
+
+            if (!s_contentHasherByHashType.ContainsKey(hashType))
+            {
+                s_contentHasherByHashType[hashType] = HashInfoLookup.Find(hashType).CreateContentHasher();
+            }
+
+            return s_contentHasherByHashType[hashType];
```
Comment (narasamdya) [inline Public/Src/Utilities/Storage/ContentHashingUtilities.cs]: Use `GetOrAdd` method of `ConcurrentDictionary`.

File: Public/Src/Engine/Dll/FrontEndEngineImplementation.cs [hunk 3]
```diff
@@ -274,13 +274,14 @@ public override async Task<ContentHash> GetFileContentHashAsync(string path, boo
 
                 VersionedFileIdentityAndContentInfo? maybeKnownIdentityAndHash = fileContentTable.TryGetKnownContentHash(fs);
 
-                if (maybeKnownIdentityAndHash.HasValue)
+                if (maybeKnownIdentityAndHash.HasValue && 
+                    (hashType == HashType.Unknown || maybeKnownIdentityAndHash.Value.FileContentInfo.Hash.HashType == hashType))
```
Comment (narasamdya) [inline Public/Src/Engine/Dll/FrontEndEngineImplementation.cs]: Perhaps in `FileContentInfo`, you add the following method:

```
bool MatchHashType(HashType hashType) => hashType == HashType.Unknown /* unknown matches everything */ || HashType == hashType;
```

I believe that method can be useful when needed in other places.

**Pull request merged**